### PR TITLE
gRelayer: make relayprovider calls safe

### PR DESCRIPTION
**Pull request by derpy-duck**

### CODE REVIEW

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 1]
```diff
@@ -326,56 +345,75 @@ abstract contract CoreRelayerDelivery is CoreRelayerGovernance {
...
+        uint256 wormholeMessageFee = wormhole().messageFee();
+        uint256 receiverValueTarget;
+        uint256 maximumBudget;
+        address rewardAddress;
+
+        if (refundAmount <= wormholeMessageFee) {
+            return (false, RefundStatus.CROSS_CHAIN_REFUND_FAIL_NOT_ENOUGH);
+        }
 
-        if (refundAmount > overhead) {
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:335]: Where are we ensuring the refund is greater than the overhead in the new version?

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 2]
```diff
@@ -223,18 +223,26 @@ abstract contract CoreRelayerDelivery is CoreRelayerGovernance {
...
         if (forwardInstructions.length > 0) {
             // If the user made a forward/multichainForward request, then try to execute it
-            stack.transactionFeeRefundAmountPostForward = emitForward(stack.transactionFeeRefundAmount, forwardInstructions);
+            stack.transactionFeeRefundAmountPostForward =
+                emitForward(stack.transactionFeeRefundAmount, forwardInstructions);
             status = DeliveryStatus.FORWARD_REQUEST_SUCCESS;
         } else {
             status = stack.callToTargetContractSucceeded
-                ? (stack.callToInstructionExecutorSucceeded ? DeliveryStatus.SUCCESS : DeliveryStatus.FORWARD_REQUEST_FAILURE)
+                ? (
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:231]: nit: I know this wasn't added in this pr, but we should not use nested ternary expressions. If more than one nesting is required, prefer an if/else chain

**Review by JoeHowarth — CHANGES_REQUESTED** [review]

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 1 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:335]: if refundAmount < overhead, then receiverValueTarget will be 0, which is checked

File: ethereum/contracts/relayer/coreRelayer/ForwardWrapper.sol [hunk 3]
```diff
@@ -68,11 +68,25 @@ contract ForwardWrapper is CoreRelayerLibrary {
...
+        uint256 deliveryOverhead = relayProvider.quoteDeliveryOverhead(targetChain);
+        rewardAddress = relayProvider.getRewardAddress();
+        maximumBudget = relayProvider.quoteMaximumBudget(targetChain);
+        uint256 srcNativeCurrencyPrice = relayProvider.quoteAssetPrice(sourceChain);
+        uint256 dstNativeCurrencyPrice = relayProvider.quoteAssetPrice(targetChain);
+        (uint16 buffer, uint16 denominator) = relayProvider.getAssetConversionBuffer(targetChain);
+        receiverValueTarget = receiverValuePlusOverhead > deliveryOverhead
+            ? (
+                (receiverValuePlusOverhead - deliveryOverhead) * srcNativeCurrencyPrice * denominator
+                    / (dstNativeCurrencyPrice * (uint256(0) + denominator + buffer))
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/coreRelayer/ForwardWrapper.sol:88]: nit: (uint256(0) + denominator + buffer) -> (uint256(denominator)+ buffer)

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 4]
```diff
@@ -326,56 +345,75 @@ abstract contract CoreRelayerDelivery is CoreRelayerGovernance {
...
 
-            pay(provider.getRewardAddress(), refundAmount - wormholeMessageFee);
+        if (!isChainSupported) {
+            return (false, RefundStatus.CROSS_CHAIN_REFUND_FAIL_PROVIDER_NOT_SUPPORTED);
+        }
 
-            return (true, isMaximumBudget ? RefundStatus.CROSS_CHAIN_REFUND_SENT_MAXIMUM_BUDGET : RefundStatus.CROSS_CHAIN_REFUND_SENT);
-            
-        } else {
+        if (receiverValueTarget == 0) {
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:367]: can you put a comment here saying that the overhead was checked in the lib?

**Review by JoeHowarth — APPROVED** [review]

File: ethereum/contracts/relayer/coreRelayer/ForwardWrapper.sol (hunk 3 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/ForwardWrapper.sol:88]: done

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 4 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:369]: done

**Review by JoeHowarth — APPROVED** [review]

**Pull request merged**